examples/Cassie/cassie_fixed_point_solver.cc

```diff
@@ -0,0 +1,165 @@
+#include "examples/Cassie/cassie_fixed_point_solver.h"
+#include "multibody/kinematic/kinematic_evaluator_set.h"
+#include "multibody/kinematic/world_point_evaluator.h"
+#include "multibody/multibody_solvers.h"
+#include "multibody/multibody_utils.h"
+#include "multibody/multipose_visualizer.h"
+#include "solvers/constraint_factory.h"
+
+#include "drake/common/text_logging.h"
+#include "drake/multibody/parsing/parser.h"
+#include "drake/solvers/solve.h"
+#include "drake/solvers/snopt_solver.h"
+
+namespace dairlib {
+using Eigen::VectorXd;
+void CassieFixedPointSolver(std::string filename, double height, double mu,
+    double min_normal_force, bool linear_friction_cone, VectorXd* q_result,
+    VectorXd* u_result, VectorXd* lambda_result, bool draw_pose) {
+  // Build plant
+  drake::multibody::MultibodyPlant<double> plant(0);
+  drake::multibody::MultibodyPlant<double> plant2(0);
+  drake::multibody::Parser parser(&plant);
+  std::string full_name = FindResourceOrThrow(filename);
+  parser.AddModelFromFile(full_name);
+  plant.mutable_gravity_field().set_gravity_vector(-9.81 *
+                                                   Eigen::Vector3d::UnitZ());
+  plant.Finalize();
+
+  multibody::KinematicEvaluatorSet<double> evaluators(plant);
+
+  // Add loop closures
+  auto left_loop = LeftLoopClosureEvaluator(plant);
+  auto right_loop = RightLoopClosureEvaluator(plant);
+  evaluators.add_evaluator(&left_loop);
+  evaluators.add_evaluator(&right_loop);
+
+  // Add contact points
+  auto left_toe = LeftToe(plant);
+  auto left_toe_evaluator = multibody::WorldPointEvaluator(plant,
+      left_toe.first, left_toe.second, Eigen::Vector3d(0,0,1),
+      Eigen::Vector3d::Zero(), false);
+  evaluators.add_evaluator(&left_toe_evaluator);
+
+  auto right_toe = RightToe(plant);
+  auto right_toe_evaluator = multibody::WorldPointEvaluator(plant,
+      right_toe.first, right_toe.second, Eigen::Vector3d(0,0,1),
+      Eigen::Vector3d::Zero(), false);
+  evaluators.add_evaluator(&right_toe_evaluator);
+
+  auto left_heel = LeftHeel(plant);
+  auto left_heel_evaluator = multibody::WorldPointEvaluator(plant,
+      left_heel.first, left_heel.second, Eigen::Vector3d(0,0,1),
+      Eigen::Vector3d::Zero(), false);
+  evaluators.add_evaluator(&left_heel_evaluator);
+
+  auto right_heel = RightHeel(plant);
+  auto right_heel_evaluator = multibody::WorldPointEvaluator(plant,
+      right_heel.first, right_heel.second, Eigen::Vector3d(0,0,1),
+      Eigen::Vector3d::Zero(), false);
+  evaluators.add_evaluator(&right_heel_evaluator);
+
+  auto program = multibody::MultibodyProgram(plant, evaluators);
+
+  auto positions_map = multibody::makeNameToPositionsMap(plant);
+  auto q = program.AddPositionVariables();
+  auto u = program.AddInputVariables();
+  auto lambda = program.AddConstraintForceVariables();
+  auto kinematic_constraint = program.AddKinematicConstraint(q);
```

Comment:
The workflow of constructing the MathematicalProgram with the kinematic evaluators, and then later calling AddKinematicConstraint() with only the position decision variables as the parameter is not easy to read. Would it be possible to adjust the API to have MultibodyProgram be constructed using only the MBP and have AddKinematicConstraint() take in both q and evaluators?

Reply:
I'm not sure I did this the right way, so I'm happy to have this discussion. Right now, the code straddles a weird place where, because it extends MathematicalProgram, it's trying to "own" the plant/evaluators/decision variables/etc. However, I don't think it's doing so particularly effectively for the reason you stated. If I remove owning the evaluators, given that ownership of the decision variables is already a fraud, then the only thing in the constructor would be plant itself. At this point, is MultibodyProgram really a class?

These could, instead, just be global utility methods.
```
DecisionVariables  q = AddPositionVariables(program*, plant&)
...
DecisionVariables lambda = AddConstraintForceVariables(program*, plant&)
Binding = AddKinematicConstraint(program*, plant&, evaluators&, q)
Binding = AddFixedPointConstraint(program*, plant&, evaluators&, q, u, lambda)
```

Reply:
I prefer keeping MultibodyProgram as a class, it has enough utility (the various AddConstraint functions that only make sense when they refer to a MBP) that it's worth keeping them together as a single class rather than as separate utility methods.

I would prefer an API that looked like:

```
DecisionVariables q = program.AddPositionVariables()
Binding = program.AddKinematicConstraint(evaluators&, q)
```

That way it's easy to see the dependencies.

Reply:
To add my point of view. I think it's pretty easy to read `MultibodyProgram`, but I admit I expected that `AddKinematicConstraint` takes in `evaluators` as an argument, the first time I saw the function. 

If we are going to have `AddKinematicConstraint(evaluators, q)`, then we will need to make sure that the users pass the same `evaluators` into both  `AddKinematicConstraint` and `AddFixedPointConstraint`. I don't have a clean solution to implement this right now

Reply:
I agree with both of you, and have a new idea that might satisfy everyone. The basic question, I think, is whether or not a `MultibodyProgram` is allowed have two evaluator sets, or multiple robot states/inputs/forces. If the answer is "no," then we can take the current implementation and _remove_ the `q,u,lambda` inputs to the `AddXXConstraint()` functions, and just DEMAND that they've been created.

However, if the answer is "yes" then the API becomes cleaner. Like Will suggested, no `evaluators` as a constructor argument. The program will no longer keep references to `q,u,lambda`, as these may be non-unique. Instead, `AddFixedPointConstraint(evaluators,q,u,lambda)` will confirm that the lengths of `q,u,lambda` are appropriate for the plant and evaluators. Similar for `AddKinematicConstraint`.

Everything becomes a bit more flexible, and the restrictions on the arguments are that 

1. evaluators.plant() matches self.plant()
2. The lengths of the input decision variables are appropriate.

Thoughts?

Reply:
From what I can imagine, we probably don't need multiple states/inputs/forces and evaluator sets. People could just create multiple `MultibodyProgram`s to achieve this?

And I have a question about the proposed structure, do we demand that the evaluators passed into `AddKinematicConstraint` and `AddFixedPointConstraint` are the same?

Reply:
This framework would flexible enough to optimize for two different states, where those states are related somehow.

Via the proposed structure, that would not be enforced--since there's really no need to.